qt6-qtbase: Fix building on MacOS 10.14 SDK

tuffnatty opened the pull request:
#### Description

As it turns out, #21831 fixed only building under macOS 10.14 with SDK 10.15. Therefore, it helped the buildbot to build qt6-qtbase on 10.15, but 10.14 binary package fails.
Here is a followup that fixes building qt6-qtbase on SDK 10.14. 

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.14.6 18G9323 x86_64
Xcode 11.7 11E801a

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [x] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

kencu commented:
the buildbot on 10.14 has not (and will not) be upgraded to the latest xcode that 10.14 can support.

so it was expected that the 10.14 buildbot would fail with the last PR.

However, most users do upgrade to the latest xcode on 10.14, like me and obviously you, as users are continually reminded to do so, and we both confirmed the last PR would work on that system.

If you want to ensure this port will build on the 10.14 buildbot, you would have to downgrade your system to the xcode, clts, and sdks supplied by the version of xcode the buildbot has available. Otherwise it really is guesswork.

If you’re looking at buildbot logs, it will only show the first error that arises…once you fix that, the next one often comes up.

So I can’t test if this PR would fix such a 10.14 system, as I’m not planning on downgrading my xcode.

What process did you use to ensure this would build on a system that matches the buildbot’s xcode/clts/sdks?

tuffnatty commented:
@kencu I have noticed the buildbot error. I then managed to reproduce it on my machine by enforcing 10.14 SDK path and clang-mp 16 usage (as the buildbot logs show clang-mp 16 is used as the fallback compiler). I have fixed the buildbot error by adding a `#if`. Then another error appeared, which was not on the failed buildbot build; so I added a patch to `src/plugins/platforms/cocoa/qcocoascreen.mm`. Then the build succeeded, locally. I am confident, that it does not prove it will build better in the buildbot environment, but I think there is a good chance.

kencu commented:
how did you enforce the sdk path exactly?

tuffnatty commented:
That's a good question. I think I added:

    macosx_deployment_target 10.14 

to `macports.conf` which had this effect.

kencu commented:
OK, that's not actually how you properly force an SDK, though.

you would do this, to force the SDK:

```
macosx_sdk_version           10.14
```

And even then, the qt build process uses an extensive shell script to set up it's build, and it will search for and find the SDK it wants to use anyway -- it can be quite tricky to force it to use another SDK than the one it wants to use, to be honest. I have been doing this with qt for years now, and fielding dozens and dozens of tickets about the wrong SDK being found. 

Point being, have to be careful, because what you think might be happening may not be happening.

When I build qt6-qtbase on 10.14 with xcode 11 (the last one that 10.14 can support), it builds first of all:
```
$ port -v installed qt6-qtbase
The following ports are currently installed:
  qt6-qtbase @6.4.3_5+openssl (active) requested_variants='' platform='darwin 18' archs='x86_64' date='2023-12-27T10:27:26-0800'
```

with the system xcode clang:
```
$ /usr/bin/clang++ -v
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin18.7.0
```
and with the 10.15 SDK, indeed:
```
-isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
```
in full:
```
Executing:  cd "/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3" && ninja -j24 all -v 
[1/3727] /opt/local/bin/ccache /usr/bin/clang++ -DGL_SILENCE_DEPRECATION -DHAVE_CONFIG_H -DOPENSSL_INCLUDE_DIR=/opt/local/libexec/openssl3/include -DOPENSSL_LIBRARIES=/opt/local/libexec/openssl3/lib -DOPENSSL_LIBS="-L/opt/local/libexec/openssl3/lib -lssl -lcrypto" -DOPENSSL_ROOT_DIR=/opt/local/libexec/openssl3 -DQT_BOOTSTRAPPED -DQT_NO_CAST_FROM_ASCII -DQT_NO_CAST_TO_ASCII -DQT_NO_DEBUG -DQT_NO_FOREACH -DQT_NO_JAVA_STYLE_ITERATORS -DQT_NO_NARROWING_CONVERSIONS_IN_CONNECT -DQT_TYPESAFE_FLAGS -DQT_USE_QSTRINGBUILDER -DQT_VERSION_MAJOR=6 -DQT_VERSION_MINOR=4 -DQT_VERSION_PATCH=3 -DQT_VERSION_STR=\"6.4.3\" -D_LARGEFILE64_SOURCE -D_LARGEFILE_SOURCE -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/corelib/Core_autogen/include -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/include -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/include/QtCore -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/lib/QtCore.framework/Headers -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/corelib -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/corelib/global -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/corelib/kernel -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/corelib/../3rdparty/tinycbor/src -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/include/QtCore/6.4.3 -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/include/QtCore/6.4.3/QtCore -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/corelib/../3rdparty/forkfd -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/corelib/.rcc -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/mkspecs/macx-clang -I/opt/local/include/glib-2.0 -I/opt/local/lib/glib-2.0/include -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/tools/bootstrap/.. -I/opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/tools/bootstrap/../../3rdparty/tinycbor/src -isystem /opt/local/include -I/opt/local/libexec/openssl3/include -DNDEBUG -Oz -arch x86_64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.14 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -ffunction-sections -fdata-sections -std=gnu++17 -MD -MT src/tools/bootstrap/CMakeFiles/Bootstrap.dir/__/__/corelib/global/qnumeric.cpp.o -MF src/tools/bootstrap/CMakeFiles/Bootstrap.dir/__/__/corelib/global/qnumeric.cpp.o.d -o src/tools/bootstrap/CMakeFiles/Bootstrap.dir/__/__/corelib/global/qnumeric.cpp.o -c /opt/local/var/macports/build/_opt_macportsofficial_macports-ports_aqua_qt6/qt6-qtbase/work/qtbase-everywhere-src-6.4.3/src/corelib/global/qnumeric.cpp
```

So if we're going to push more PRs, we have to be sure that it is worthwhile to do so.

Can you see exactly which SDK is being used in your build (from the build line)?

kencu commented:
If you want to force both the SDK and the deployment target, you'd add these two lines to macports.conf

```
macosx_sdk_version           10.14
macosx_deployment_target 10.14 
```

you might be intrigued by this ticket, where we went through the process of forcing macports do jump through all kinds of SDK and deployment target hoops about five years ago, building wine:

https://trac.macports.org/ticket/56991

It's applicable to what is happening here, but not identical.

MacPorts has been working on a process whereby a port could request a certain non-default SDK to be available for the build ... eg you could have qt6-qtbase on 10.14 say it needs the the 10.15 SDK to be available for the build, and it would magically show up somewhere.

Unfortunately, due to licensing issues (MacPorts will never be able to distribute the SDKs legally) that will probably never happen.

tuffnatty commented:
Thank you! Well, my build logs show:
```
/opt/local/var/macports/logs/_Users_phil_common-home_build_macports-ports_aqua_qt6/qt6-qtbase/main.log                           2357/189K  UTF-8         1%
version:1
:debug:main Starting logging for qt6-qtbase @6.4.3_5+openssl
:debug:sysinfo macOS 10.14.6 (darwin/18.7.0) arch i386
:debug:sysinfo MacPorts 2.8.1
:debug:sysinfo Xcode 11.7, CLT 10.3.0.0.1.1562985497
:debug:sysinfo SDK 10.14
:debug:sysinfo MACOSX_DEPLOYMENT_TARGET: 10.14
....
/opt/local/bin/ccache /opt/local/bin/clang++-mp-16 .... -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk -mmacosx-version-min=10.14 ....
```
Sure I will now try to add `macosx_sdk_version           10.14` and see if anything changes. But I am pretty sure that 1) SDK 10.14 is being used and 2) for Qt 6.4's cmake-based build, passing deployment target is enough to enforce SDK version.

tuffnatty commented:
Adding `macosx_sdk_version           10.14` does not change configure and build output on my machine.

tuffnatty commented:
In fact, removing any or both of the lines from macports.conf, or switching (using `xcode-select -s`) between several Developer folders I have on my machine, don't make the qt6-qtbase build use SDK 10.15 again, it always uses 10.14... So I take back my words on how I did achieve it, it was 3:00am.

kencu commented:
heh… making qt builds obey your bidding can be tricky, for sure.

It looks like you did force the 10.14 sdk in your build, and these changes also seem benign, so even though I can’t test it (easily) it should work.

Thanks again for your time and efforts.
